### PR TITLE
Get endpoint by  Target id implemented

### PR DESCRIPTION
**Pull request by humulla**
## Link to JIRA ticket

https://hackney.atlassian.net/jira/software/projects/MTTL/boards/47?selectedIssue=MTTL-491

### The following tasks have been completed
- [x] Infrastructure set up 
- [x] Controller set up
- [x] Gateway set up
- [x] UseCase set up  
- [x] Domain set up
- [x] Factories (domain and response) set up
- [x] Boundary Response and Request set up 
- [x] Added end-to-end (i.e. integration) tests where necessary e.g to test the complete functionality of a newly added feature
- [x] Added tests to cover all new production code
- [x] Checked all code for possible refactoring
- [x] Code pipeline builds correctly

### CODE REVIEW

File: ContactDetailsApi/V1/Infrastructure/DatabaseContext.cs [hunk 1]
```diff
@@ -11,6 +11,6 @@ public DatabaseContext(DbContextOptions options) : base(options)
         {
         }
 
-        public DbSet<DatabaseEntity> DatabaseEntities { get; set; }
+        public DbSet<ContactDetailsEntity> DatabaseEntities { get; set; }
```
Comment (rocco-scaramuzzi) [inline ContactDetailsApi/V1/Infrastructure/DatabaseContext.cs:14]: I think we can remove dbContext as we are not using SQL

**Review by rocco-scaramuzzi — CHANGES_REQUESTED** [review]

File: ContactDetailsApi/V1/Controllers/ContactDetailsController.cs [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+        /// <response code="500">Internal Server Error</response>
+        [ProducesResponseType(typeof(ContactDetailsResponseObject), StatusCodes.Status200OK)]
+        [ProducesResponseType(StatusCodes.Status404NotFound)]
+        [ProducesResponseType(StatusCodes.Status500InternalServerError)]
+        [HttpGet]
+        public async Task<IActionResult> GetContactByTargetId([FromQuery] ContactQueryParameter cqp)
+        {
+
+            var contact = await _getContactByTargetIdUseCase.Execute(cqp).ConfigureAwait(false);
+            if (null == contact || contact.Count == 0) return NotFound(cqp);
```
[the remark this answers is not in the document]
Comment (mikemoody-amido) [reply ContactDetailsApi/V1/Controllers/ContactDetailsController.cs:39]: > always have the variable you want to check at the left side of the operator (contacts == null)
I always do the opposite, but then again I originally trained in C++ where inadvertently missing out one of the '=' could cause massive problems...

File: ContactDetailsApi/V1/Domain/InvalidQueryParameterException.cs [hunk 3]
```diff
@@ -0,0 +1,15 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace ContactDetailsApi.V1.Domain
+{
+    public class InvalidQueryParameterException : Exception
```
Comment (mikemoody-amido) [inline ContactDetailsApi/V1/Domain/InvalidQueryParameterException.cs:8]: Is this exception class used anywhere?

File: ContactDetailsApi.Tests/V1/UseCase/GetByIdUseCaseTests.cs [hunk 4]
```diff
@@ -1,23 +1,78 @@
...
 using Moq;
 using NUnit.Framework;
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
 
 namespace ContactDetailsApi.Tests.V1.UseCase
 {
+    [TestFixture]
     public class GetByIdUseCaseTests
```
Comment (mikemoody-amido) [inline ContactDetailsApi.Tests/V1/UseCase/GetByIdUseCaseTests.cs:18]: Should rename this (and the filename) to reflect the changed use case class name.

File: ContactDetailsApi.Tests/V1/E2ETests/E2EDynamoDbTest.cs [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+        /// <param name="entity"></param>
+        /// <returns></returns>
+        private async Task SetupTestData(ContactDetails entity)
+        {
+            await DynamoDbContext.SaveAsync(entity.ToDatabase()).ConfigureAwait(false);
+            CleanupActions.Add(async () => await DynamoDbContext.DeleteAsync<ContactDetailsEntity>(entity.TargetId, entity.Id).ConfigureAwait(false));
+        }
+
+        [Test]
+        public async Task GetEntityByIdNotFoundReturns404()
```
Comment (mikemoody-amido) [inline ContactDetailsApi.Tests/V1/E2ETests/E2EDynamoDbTest.cs:46]: Should update the name to reflect Contact details.

File: ContactDetailsApi.Tests/V1/E2ETests/E2EDynamoDbTest.cs [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+        {
+            var targetId = Guid.NewGuid().ToString();
+            var uri = new Uri($"api/v1/contactDetails?targetId={targetId}", UriKind.Relative);
+            var response = await Client.GetAsync(uri).ConfigureAwait(false);
+
+            response.StatusCode.Should().Be(HttpStatusCode.NotFound);
+        }
+
+        [Test]
+        public async Task GetNoteBydIdFoundReturnsResponse()
```
Comment (mikemoody-amido) [inline ContactDetailsApi.Tests/V1/E2ETests/E2EDynamoDbTest.cs:56]: Should update the name to reflect Contact details.

File: ContactDetailsApi/V1/Controllers/ContactDetailsController.cs [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+
+namespace ContactDetailsApi.V1.Controllers
+{
+    [ApiController]
+    [Route("api/v1/contactDetails")]
+    [Produces("application/json")]
+    [ApiVersion("1.0")]
+    public class ContactDetailsController : BaseController
+    {
+        private readonly IGetByIdUseCase _getByIdUseCase;
```
[the remark this answers is not in the document]
Comment (humulla) [reply ContactDetailsApi/V1/Controllers/ContactDetailsController.cs]: This is an old commit, i've already changed the filename to GetContactByTargetIdUseCase

File: ContactDetailsApi/V1/Controllers/ContactDetailsController.cs [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+        /// Retrieves contact details for a particular id
+        /// </summary>
+        /// <response code="200">Successfully Operation</response>
+        /// <response code="404">Contact not found for ID requested</response>
+        /// <response code="500">Internal Server Error</response>
+        [ProducesResponseType(typeof(ContactDetailsResponseObject), StatusCodes.Status200OK)]
+        [ProducesResponseType(StatusCodes.Status404NotFound)]
+        [ProducesResponseType(StatusCodes.Status500InternalServerError)]
+        [HttpGet]
+        [Route("{id}")]
```
[the remark this answers is not in the document]
Comment (humulla) [reply ContactDetailsApi/V1/Controllers/ContactDetailsController.cs]: This is an old commit, i've already removed this after our three amigo session

File: ContactDetailsApi/V1/Boundary/Response/ContactDetailsResponseObject.cs [hunk 9]
```diff
@@ -0,0 +1,18 @@
+using ContactDetailsApi.V1.Domain;
+using System;
+using System.Collections.Generic;
+
+namespace ContactDetailsApi.V1.Boundary.Response
+{
+    public class ContactDetailsResponseObject
+    {
```
[the remark this answers is not in the document]
Comment (humulla) [reply ContactDetailsApi/V1/Boundary/Response/ContactDetailsResponseObject.cs:8]: I've changed them, I understood the Swagger doc incorrectly

[the remark this answers is not in the document]
Comment (humulla) [reply ContactDetailsApi/V1/Boundary/Response/ContactDetailsResponseObject.cs:8]: I've removed it, I think i misinterpreted the swagger doc

File: ContactDetailsApi/V1/Infrastructure/ContactDetailsEntity.cs [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+        public List<SourceServiceArea> SourceServiceArea { get; set; }
+
+        [DynamoDBProperty(Converter = typeof(DynamoDbDateTimeConverter))]
+        public DateTime RecordValidUntil { get; set; }
+
+        public bool IsActive { get; set; }
+
+        [DynamoDBProperty(Converter = typeof(DynamoDbObjectListConverter<CreatedBy>))]
+
+        public List<CreatedBy> CreatedBy { get; set; }
```
[the remark this answers is not in the document]
Comment (humulla) [reply ContactDetailsApi/V1/Infrastructure/ContactDetailsEntity.cs:38]: I've removed them

File: ContactDetailsApi/V1/Domain/InvalidQueryParameterException.cs (hunk 3 above)
Comment (humulla) [reply ContactDetailsApi/V1/Domain/InvalidQueryParameterException.cs:8]: I thought i would use it, but I didn't, have deleted it

**Review by mikemoody-amido — APPROVED** [review]

**Pull request merged**